lib/graphql/dataloader/loader.rb

```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+module GraphQL
+  class Dataloader
+    class Loader
+      class StateError < GraphQL::Error; end
+
+      def self.load(context, key, value)
+        dl = context[:dataloader]
+        loader = dl.loaders[self][key]
+        loader.load(value)
+      end
+
+      def initialize(context, key)
+        @context = context
+        @key = key
+        @promises = {}
+        @loaded_values = {}
+      end
+
+      def load(value)
+        @promises[value] ||= GraphQL::Execution::Lazy.new do
+          if !@loaded_values.key?(value)
+            sync
+          end
+          @loaded_values[value]
+        end
+      end
+
+      def sync
+        # Promises might be added in the meantime, but they won't be included in this list.
+        keys_to_load = @promises.keys - @loaded_values.keys
+        resolved_values = perform(keys_to_load)
+        keys_to_load.each_with_index do |k, i|
+          resolved_value = resolved_values[i]
+          @loaded_values[k] = resolved_value
+        end
+        nil
+      end
+
+      def perform(values)
+        raise NotImplementedError, "`#{self.class}#perform` should load `values` for `@key` and return an item for each of `values`"
```

Comment:
graphql-batch has a separate method for fulfilling rather than relying on an array return value; not sure which one I prefer "from scratch", but not really looking forward to migrating :P

Reply:
:+1: I think our base loader implements this API on top of that one. I'll make sure to retain both, or something.

Reply:
A return array is probably easier to reason about in simple cases, but seems like you'd run into some hairiness when `values` (should that be `keys`?) contains duplicates, or there are a lot of keys that don't have a record in the DB, etc.

Reply:
I may not fully understand this implementation, but this does not seem to allow for multiple background loaders to be running IO at the same time. Do you not need to wrap the `perform` call in some sort of future/thread as well?

Reply:
I switched to a fulfill-based API like graphql-batch has, thanks again for taking a look. 

I think you're right that the Future is in the wrong place. I'll take a closer look at that soon.